File: JavaSource/src/main/java/ru/akhcheck/patterns/command/Waiter.java
package ru.akhcheck.patterns.command;

import java.util.ArrayList;
import java.util.List;
import ru.akhcheck.patterns.command.commands.Command;

/**
 * Тот, кто принимает заказ, но не готовит.
 *
 * <p>Официант знает только про интерфейс команды. Кто исполнитель и что именно
 * он сделает, официанту неизвестно, и в этом весь смысл: добавление нового
 * блюда не меняет здесь ни строчки.
 *
 * <p>Счёт и отмена возможны ровно потому, что команды это объекты, которые
 * можно сохранить.
 */
public class Waiter {
    private final List<Command> commands = new ArrayList<>();

    public void setCommand(Command command) {
        commands.add(command);
    }

    public void executeCommands() {
        for (Command command : commands) {
            System.out.println("Order: " + command);
            command.execute();
        }
    }

    public int getBill() {
        int sum = 0;
        for (Command command : commands) {
            sum += command.getPrice();
        }
        return sum;
    }

    public Command cancelLast() {
        if (commands.isEmpty()) {
            return null;
        }
        return commands.remove(commands.size() - 1);
    }
}

File: Command/python-source/Waiter.py
from __future__ import annotations

from commands import Command


class Waiter:
    """Тот, кто принимает заказ, но не готовит.

    Официант знает только про интерфейс команды. Кто исполнитель и что именно
    он сделает, официанту неизвестно, и в этом весь смысл: добавление нового
    блюда не меняет ни строчки здесь.

    Заодно официант ведёт счёт, и он возможен только потому, что команды это
    объекты, которые можно сохранить.
    """

    def __init__(self) -> None:
        self._commands: list[Command] = []

    def set_command(self, command: Command) -> None:
        self._commands.append(command)

    def execute_commands(self) -> None:
        for command in self._commands:
            print(f"Order: {command}")
            command.execute()

    def get_bill(self) -> int:
        return sum(command.get_price() for command in self._commands)

    def cancel_last(self) -> Command | None:
        """Отмена возможна ровно потому, что запрос стал объектом."""
        if not self._commands:
            return None
        return self._commands.pop()
